### ISSUE TITLE
datetimes and dates are not substitutable but ty thinks they are

### ISSUE THREAD

**Issue by m-aciek** [astral-sh/ty#615]
```python
from datetime import date, datetime

if datetime.now() < date.today():
    print("that's a surprise!")
```

* What is the actual behavior/output?
No error!

* What is the behavior/output you expect?
A warning, since at runtime we get `TypeError: can't compare datetime.datetime to datetime.date`.

`datetime` subclassing `date` is a violation of Liskov substitution principle in Python. Though I'd expect typecheckers to at least warn the developer.

Related report in mypy: https://github.com/python/mypy/issues/9015.

**Comment by AlexWaygood** [comment]
Thanks for opening the issue. No type checker that I'm aware of issues a complaint about this code, so I would classify this as a feature rather than a bug. As you say, the fundamental problem is that `datetime` breaks the Liskov Substitution Principle.

It's possible we could add some logic like Jukka describes in https://github.com/python/mypy/issues/9015#issuecomment-646678572, but this definitely wouldn't be a priority for us since it goes above and beyond compliance with the typing spec. It also seems like the kind of thing that could possibly be expensive to check fully.

Another option would, I suppose, be to hardcode some special-cased logic in specifically for `datetime` and `date`. I don't like that option though: adding special-casing like that doesn't scale well (several other classes in the CPython stdlib are unsound in the same way), creates inconsistencies that can be confusing to users, and is prone to breakage.

**Comment by AlexWaygood** [comment]
(In the meantime, [as you know](https://github.com/python/mypy/issues/9015#issuecomment-1621406542), another option is to use a more modern datetime library than the CPython stdlib one, such as [`datetype`](https://github.com/glyph/DateType) or [`whenever`](https://github.com/ariebovenberg/whenever))

**Comment by MichaReiser** [comment]
Somewhat related to https://github.com/astral-sh/ty/issues/576

**Comment by AlexWaygood** [comment]
> Somewhat related to [#576](https://github.com/astral-sh/ty/issues/576)

eh, not really:
- strict-equality is about catching expressions that likely always evaluate to `False` or `True` (probably unintentionally). This issue is about catching exceptions that arise from non-equality comparisons. It's debatable whether strict-equality should even be a type checker issue or whether it should be a linter issue; this issue is definitely a type-checking issue.
- strict-equality is something that existing type checkers already have some kind of solution for (even though all existing solutions use somewhat ugly heuristics). This is something that no type checker has a solution for really.
- This issue only arises because the CPython stdlib is breaking some fundamental rules of subtyping, which typeshed is forced to mimic in its stubs, and this breaks some fundamental assumptions made by ty. (You can see this from the number of `type: ignore[override]` comments in [typeshed's stubs for the datetime module](https://github.com/python/typeshed/blob/8f5a80e7b741226e525bd90e45496b8f68dc69b6/stdlib/datetime.pyi#L331-L334)!) None of that's the case for strict-equality considerations; strict-equality isn't really an issue to do with type safety at all.

**Comment by m-aciek** [comment]
I've opened a fix covering operators at https://github.com/astral-sh/ruff/pull/28771. I've discussed this topic with Jukka Lehtosalo on EuroPython sprints in July this year, and he wasn't keen to reconsider changing the typeshed, and confirmed that call-site discovery in his opinion is the right approach.

~~Please note that this relation is the only LSP violation in the standard library,~~ in my opinion it deserves special casing. cc @sharkdp

**Comment by AlexWaygood** [comment]
> Please note that this relation is the only LSP violation in the standard library

I wish that were true, but it certainly isn't

<details>
<summary>List of stdlib locations where typeshed has `type: ignore` comments to suppress Liskov violations</summary>

```
_ctypes.pyi:    def __ctypes_from_outparam__(self, /) -> _T: ...  # type: ignore[override]
_hashlib.pyi:    def digest(self, length: int) -> bytes: ...  # type: ignore[override]
_hashlib.pyi:    def hexdigest(self, length: int) -> str: ...  # type: ignore[override]
_interpchannels.pyi:    def count(self) -> int: ...  # type: ignore[override]
_io.pyi:    def __iter__(self) -> Iterator[str]: ...  # type: ignore[override]
_io.pyi:    def __next__(self) -> str: ...  # type: ignore[override]
_io.pyi:    def writelines(self, lines: Iterable[str], /) -> None: ...  # type: ignore[override]
_io.pyi:    def readline(self, size: int = -1, /) -> str: ...  # type: ignore[override]
_io.pyi:    def readlines(self, hint: int = -1, /) -> list[str]: ...  # type: ignore[override]
_io.pyi:    def buffer(self) -> _BufferT_co: ...  # type: ignore[override]
_io.pyi:    def readline(self, size: int = -1, /) -> str: ...  # type: ignore[override]
_io.pyi:    def detach(self) -> _BufferT_co: ...  # type: ignore[override]
array.pyi:    @overload  # type: ignore[override]
array.pyi:    def __iadd__(self, value: array[_T], /) -> Self: ...  # type: ignore[override]
ast.pyi:            def __replace__(  # type: ignore[override]
asyncio/base_events.pyi:    def set_default_executor(self, executor: ThreadPoolExecutor) -> None: ...  # type: ignore[override]
asyncio/base_events.pyi:        async def create_datagram_endpoint(  # type: ignore[override]
asyncio/base_subprocess.pyi:    def get_pid(self) -> int | None: ...  # type: ignore[override]
asyncio/base_subprocess.pyi:    def get_pipe_transport(self, fd: int) -> _File: ...  # type: ignore[override]
asyncio/protocols.pyi:    def connection_made(self, transport: transports.DatagramTransport) -> None: ...  # type: ignore[override]
builtins.pyi:    def __contains__(self, key: str, /) -> bool: ...  # type: ignore[override]
builtins.pyi:    def __contains__(self, key: SupportsIndex | ReadableBuffer, /) -> bool: ...  # type: ignore[override]
builtins.pyi:    def __iadd__(self, value: ReadableBuffer, /) -> Self: ...  # type: ignore[override]
builtins.pyi:    def __contains__(self, key: SupportsIndex | ReadableBuffer, /) -> bool: ...  # type: ignore[override]
builtins.pyi:    @overload  # type: ignore[override]
builtins.pyi:        @overload  # type: ignore[override]
builtins.pyi:    def index(self, value: int, /) -> int: ...  # type: ignore[override]
builtins.pyi:        @overload  # type: ignore[override]
builtins.pyi:        @overload  # type: ignore[override]
bz2.pyi:    def readline(self, size: SupportsIndex = -1) -> bytes: ...  # type: ignore[override]
codecs.pyi:    def write(self, data: str) -> None: ...  # type: ignore[override]
codecs.pyi:    def seek(self, offset: int, whence: int = 0) -> None: ...  # type: ignore[override]
codecs.pyi:    def write(self, data: bytes) -> None: ...  # type: ignore[override]
codecs.pyi:    def writelines(self, list: Iterable[bytes]) -> None: ...  # type: ignore[override]
codecs.pyi:    def seek(self, offset: int, whence: int = 0) -> None: ...  # type: ignore[override]
collections/__init__.pyi:    def pop(self) -> _T: ...  # type: ignore[override]
collections/__init__.pyi:    def __getitem__(self, key: SupportsIndex, /) -> _T: ...  # type: ignore[override]
collections/__init__.pyi:    def __setitem__(self, key: SupportsIndex, value: _T, /) -> None: ...  # type: ignore[override]
collections/__init__.pyi:    def __delitem__(self, key: SupportsIndex, /) -> None: ...  # type: ignore[override]
collections/__init__.pyi:    def fromkeys(cls, iterable: Any, v: int | None = None) -> Never: ...  # type: ignore[override]
collections/__init__.pyi:    @overload  # type: ignore[override]
collections/__init__.pyi:    def __or__(self, other: Counter[_S]) -> Counter[_T | _S]: ...  # type: ignore[override]
collections/__init__.pyi:        def __xor__(self, other: Counter[_S]) -> Counter[_T | _S]: ...  # type: ignore[override]
collections/__init__.pyi:        @overload  # type: ignore[override]
collections/__init__.pyi:    @overload  # type: ignore[override]
collections/__init__.pyi:    @overload  # type: ignore[override]
compression/zstd/_zstdfile.pyi:    def flush(self, mode: _ZstdCompressorFlushBlock | _ZstdCompressorFlushFrame = 1) -> bytes: ...  # type: ignore[override]
concurrent/futures/interpreter.pyi:        @overload  # type: ignore[override]
concurrent/futures/interpreter.pyi:        def run(self, task: _Task) -> None: ...  # type: ignore[override]
concurrent/futures/interpreter.pyi:        @overload  # type: ignore[override]
configparser.pyi:    @overload  # type: ignore[override]
configparser.pyi:    @overload  # type: ignore[override]
configparser.pyi:    @overload  # type: ignore[override]
datetime.pyi:    def __le__(self, value: datetime, /) -> bool: ...  # type: ignore[override]
datetime.pyi:    def __lt__(self, value: datetime, /) -> bool: ...  # type: ignore[override]
datetime.pyi:    def __ge__(self, value: datetime, /) -> bool: ...  # type: ignore[override]
datetime.pyi:    def __gt__(self, value: datetime, /) -> bool: ...  # type: ignore[override]
datetime.pyi:    @overload  # type: ignore[override]
dbm/sqlite3.pyi:    def keys(self) -> list[bytes]: ...  # type: ignore[override]
email/_header_value_parser.pyi:    def __getnewargs__(self) -> tuple[str, str]: ...  # type: ignore[override]
email/_policybase.pyi:    def header_fetch_parse(self, name: str, value: str) -> str | Header: ...  # type: ignore[override]
email/feedparser.pyi:    def feed(self, data: bytes | bytearray) -> None: ...  # type: ignore[override]
email/message.pyi:    def attach(self, payload: Self) -> None: ...  # type: ignore[override]
encodings/ascii.pyi:    def encode(data: ReadableBuffer, errors: str | None = None, /) -> tuple[str, int]: ...  # type: ignore[override]
encodings/ascii.pyi:    def decode(str: str, errors: str | None = None, /) -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/base64_codec.pyi:    def encode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/base64_codec.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/base64_codec.pyi:    def encode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/base64_codec.pyi:    def decode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/bz2_codec.pyi:    def encode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/bz2_codec.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/bz2_codec.pyi:    def encode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/bz2_codec.pyi:    def decode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/charmap.pyi:    def encode(self, input: str, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/charmap.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[str, int]: ...  # type: ignore[override]
encodings/hex_codec.pyi:    def encode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/hex_codec.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/hex_codec.pyi:    def encode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/hex_codec.pyi:    def decode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/latin_1.pyi:    def encode(data: ReadableBuffer, errors: str | None = None, /) -> tuple[str, int]: ...  # type: ignore[override]
encodings/latin_1.pyi:    def decode(str: str, errors: str | None = None, /) -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/punycode.pyi:    def decode(self, input: memoryview | bytes | bytearray | str, final: bool = False) -> str: ...  # type: ignore[override]
encodings/quopri_codec.pyi:    def encode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/quopri_codec.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/quopri_codec.pyi:    def encode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/quopri_codec.pyi:    def decode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/raw_unicode_escape.pyi:    def decode(self, input: str | ReadableBuffer, errors: str = "strict") -> tuple[str, int]: ...  # type: ignore[override]
encodings/rot_13.pyi:    def encode(self, input: str, errors: str = "strict") -> tuple[str, int]: ...  # type: ignore[override]
encodings/rot_13.pyi:    def decode(self, input: str, errors: str = "strict") -> tuple[str, int]: ...  # type: ignore[override]
encodings/rot_13.pyi:    def encode(self, input: str, final: bool = False) -> str: ...  # type: ignore[override]
encodings/rot_13.pyi:    def decode(self, input: str, final: bool = False) -> str: ...  # type: ignore[override]
encodings/unicode_escape.pyi:    def decode(self, input: str | ReadableBuffer, errors: str = "strict") -> tuple[str, int]: ...  # type: ignore[override]
encodings/utf_8_sig.pyi:    def setstate(self, state: int) -> None: ...  # type: ignore[override]
encodings/uu_codec.pyi:    def encode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/uu_codec.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/uu_codec.pyi:    def encode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/uu_codec.pyi:    def decode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/zlib_codec.pyi:    def encode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/zlib_codec.pyi:    def decode(self, input: ReadableBuffer, errors: str = "strict") -> tuple[bytes, int]: ...  # type: ignore[override]
encodings/zlib_codec.pyi:    def encode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
encodings/zlib_codec.pyi:    def decode(self, input: ReadableBuffer, final: bool = False) -> bytes: ...  # type: ignore[override]
enum.pyi:        @overload  # type: ignore[override]
enum.pyi:    def __prepare__(metacls, cls: str, bases: tuple[type, ...], **kwds: Any) -> _EnumDict: ...  # type: ignore[override]
enum.pyi:    def name(self) -> str | None: ...  # type: ignore[override]
fractions.pyi:    def __hash__(self) -> int: ...  # type: ignore[override]
http/client.pyi:    def readline(self, limit: int = -1) -> bytes: ...  # type: ignore[override]
http/cookies.pyi:    def update(self, values: Iterable[tuple[str, str]] | SupportsKeysAndGetItem[str, str]) -> None: ...  # type: ignore[override]
importlib/metadata/__init__.pyi:    def name(self) -> str: ...  # type: ignore[override]
importlib/metadata/__init__.pyi:        def __getitem__(self, name: str) -> EntryPoint: ...  # type: ignore[override]
importlib/metadata/__init__.pyi:        def __getitem__(self, name: int | str) -> EntryPoint: ...  # type: ignore[override]
importlib/readers.pyi:    def read_text(self, *args: Never, **kwargs: Never) -> Never: ...  # type: ignore[override]
importlib/readers.pyi:        def joinpath(self, child: StrPath) -> abc.Traversable: ...  # type: ignore[override]
importlib/readers.pyi:    def open(self, *args: Never, **kwargs: Never) -> Never: ...  # type: ignore[override]
importlib/resources/simple.pyi:        def joinpath(self, name: Never) -> Never: ...  # type: ignore[override]
importlib/resources/simple.pyi:        def open(self, *args: Never, **kwargs: Never) -> Never: ...  # type: ignore[override]
lib2to3/main.pyi:    def log_error(  # type: ignore[override]
lib2to3/main.pyi:    def write_file(  # type: ignore[override]
lib2to3/main.pyi:    def print_output(self, old: str, new: str, filename: str, equal: bool) -> None: ...  # type: ignore[override]
lib2to3/pgen2/pgen.pyi:    def __eq__(self, other: DFAState) -> bool: ...  # type: ignore[override]
logging/handlers.pyi:    def makeSocket(self) -> socket: ...  # type: ignore[override]
multiprocessing/managers.pyi:        @overload  # type: ignore[override]
multiprocessing/managers.pyi:        def keys(self) -> list[_KT]: ...  # type: ignore[override]
multiprocessing/managers.pyi:        def items(self) -> list[tuple[_KT, _VT]]: ...  # type: ignore[override]
multiprocessing/managers.pyi:        def values(self) -> list[_VT]: ...  # type: ignore[override]
multiprocessing/managers.pyi:        @overload  # type: ignore[override]
multiprocessing/managers.pyi:        def keys(self) -> list[_KT]: ...  # type: ignore[override]
multiprocessing/managers.pyi:        def items(self) -> list[tuple[_KT, _VT]]: ...  # type: ignore[override]
multiprocessing/managers.pyi:        def values(self) -> list[_VT]: ...  # type: ignore[override]
multiprocessing/managers.pyi:    def __iadd__(self, value: Iterable[_T], /) -> Self: ...  # type: ignore[override]
multiprocessing/managers.pyi:    def __imul__(self, value: SupportsIndex, /) -> Self: ...  # type: ignore[override]
multiprocessing/queues.pyi:    def __getstate__(self) -> _JoinableQueueState: ...  # type: ignore[override]
multiprocessing/queues.pyi:    def __setstate__(self, state: _JoinableQueueState) -> None: ...  # type: ignore[override]
multiprocessing/sharedctypes.pyi:    @overload  # type: ignore[override]
multiprocessing/sharedctypes.pyi:    @overload  # type: ignore[override]
multiprocessing/sharedctypes.pyi:    def __getslice__(self, start: SupportsIndex, stop: SupportsIndex) -> bytes: ...  # type: ignore[override]
multiprocessing/sharedctypes.pyi:    def __setslice__(self, start: SupportsIndex, stop: SupportsIndex, values: bytes) -> None: ...  # type: ignore[override]
pdb.pyi:        def completenames(self, text: str, line: str, begidx: int, endidx: int) -> list[str]: ...  # type: ignore[override]
pydoc.pyi:        def docroutine(  # type: ignore[override]
pydoc.pyi:        def docroutine(  # type: ignore[override]
pydoc.pyi:        def docproperty(self, object: object, name: str | None = None, mod: str | None = None, cl: Any | None = None) -> str: ...  # type: ignore[override]
pydoc.pyi:        def docdata(self, object: object, name: str | None = None, mod: Any | None = None, cl: Any | None = None) -> str: ...  # type: ignore[override]
pydoc.pyi:        def docroutine(  # type: ignore[override]
pydoc.pyi:        def docroutine(self, object: object, name: str | None = None, mod: str | None = None, cl: Any | None = None) -> str: ...  # type: ignore[override]
pydoc.pyi:        def docmodule(self, object: object, name: str | None = None, mod: Any | None = None) -> str: ...  # type: ignore[override]
pydoc.pyi:        def docproperty(self, object: object, name: str | None = None, mod: Any | None = None, cl: Any | None = None) -> str: ...  # type: ignore[override]
pydoc.pyi:        def docdata(self, object: object, name: str | None = None, mod: str | None = None, cl: Any | None = None) -> str: ...  # type: ignore[override]
pydoc.pyi:        def docother(  # type: ignore[override]
random.pyi:    def seed(self, a: int | float | str | bytes | bytearray | None = None, version: int = 2) -> None: ...  # type: ignore[override]  # noqa: Y041
shelve.pyi:    @overload  # type: ignore[override]
shelve.pyi:    def __contains__(self, key: str) -> bool: ...  # type: ignore[override]
smtpd.pyi:    def push(self, msg: str) -> None: ...  # type: ignore[override]
smtpd.pyi:    def process_message(self, peer: _Address, mailfrom: str, rcpttos: list[str], data: bytes | str) -> str | None: ...  # type: ignore[override]
smtpd.pyi:        def process_message(self, peer: _Address, mailfrom: str, rcpttos: list[str], data: bytes | str) -> str | None: ...  # type: ignore[override]
socketserver.pyi:    def get_request(self) -> tuple[tuple[bytes, _socket], _RetAddress]: ...  # type: ignore[override]
ssl.pyi:    def recvmsg(self, *args: Never, **kwargs: Never) -> Never: ...  # type: ignore[override]
ssl.pyi:    def recvmsg_into(self, *args: Never, **kwargs: Never) -> Never: ...  # type: ignore[override]
ssl.pyi:    def sendmsg(self, *args: Never, **kwargs: Never) -> Never: ...  # type: ignore[override]
ssl.pyi:    def protocol(self) -> _SSLMethod: ...  # type: ignore[override]
tempfile.pyi:    def readline(self, limit: int | None = ..., /) -> AnyStr: ...  # type: ignore[override]
tempfile.pyi:    def readlines(self, hint: int = ..., /) -> list[AnyStr]: ...  # type: ignore[override]
tempfile.pyi:        def truncate(self, size: int | None = None) -> None: ...  # type: ignore[override]
tempfile.pyi:    @overload  # type: ignore[override]
tempfile.pyi:    def __iter__(self) -> Iterator[AnyStr]: ...  # type: ignore[override]
tempfile.pyi:    def __next__(self) -> AnyStr: ...  # type: ignore[override]
tkinter/__init__.pyi:    def bbox(self, *args: str | int) -> tuple[int, int, int, int]: ...  # type: ignore[override]
tkinter/__init__.pyi:    def lower(self, first: str | int, second: str | int | None = ..., /) -> None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def tkraise(self, first: str | int, second: str | int | None = ..., /) -> None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def lift(self, first: str | int, second: str | int | None = ..., /) -> None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def selection_clear(self) -> None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def bbox(self, index: str | int) -> tuple[int, int, int, int] | None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def selection_clear(self, first: str | int, last: str | int | None = None) -> None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def size(self) -> int: ...  # type: ignore[override]
tkinter/__init__.pyi:    def bbox(self, index: str | float | _tkinter.Tcl_Obj | Widget) -> tuple[int, int, int, int] | None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def bbox(self, index) -> tuple[int, int, int, int] | None: ...  # type: ignore[override]
tkinter/__init__.pyi:    def selection_clear(self): ...  # type: ignore[override]
tkinter/tix.pyi:    def forget(self, name: str) -> None: ...  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    def bbox(self, index) -> tuple[int, int, int, int]: ...  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    def forget(self, tab_id) -> None: ...  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    def bbox(self, item: str | int, column: str | int | None = None) -> tuple[int, int, int, int] | Literal[""]: ...  # type: ignore[override]
tkinter/ttk.pyi:    @overload  # type: ignore[override]
tkinter/ttk.pyi:    def identify(self, component, x, y): ...  # type: ignore[override]
tracemalloc.pyi:    def __contains__(self, frame: Frame) -> bool: ...  # type: ignore[override]
turtle.pyi:        def __add__(self, other: tuple[float, float]) -> Vec2D: ...  # type: ignore[override]
turtle.pyi:        @overload  # type: ignore[override]
turtle.pyi:        def __rmul__(self, other: float) -> Vec2D: ...  # type: ignore[override]
turtle.pyi:        def __add__(self, other: tuple[float, float]) -> Vec2D: ...  # type: ignore[override]
turtle.pyi:        @overload  # type: ignore[override]
turtle.pyi:        def __rmul__(self, other: float) -> Vec2D: ...  # type: ignore[override]
types.pyi:    def __dict__(self) -> dict[str, Any]: ...  # type: ignore[override]
types.pyi:        def update(self, object: SupportsKeysAndGetItem[str, Any] | Iterable[tuple[str, Any]], /) -> None: ...  # type: ignore[override]
typing.pyi:    def __contains__(self, item: tuple[object, object], /) -> bool: ...  # type: ignore[override]
typing.pyi:        def __module__(self) -> str | None: ...  # type: ignore[override]
typing_extensions.pyi:        def __module__(self) -> str | None: ...  # type: ignore[override]
urllib/error.pyi:    def reason(self) -> str: ...  # type: ignore[override]
weakref.pyi:    def keys(self) -> Iterator[_KT]: ...  # type: ignore[override]
weakref.pyi:    def values(self) -> Iterator[_VT]: ...  # type: ignore[override]
weakref.pyi:    def items(self) -> Iterator[tuple[_KT, _VT]]: ...  # type: ignore[override]
weakref.pyi:    def keys(self) -> Iterator[_KT]: ...  # type: ignore[override]
weakref.pyi:    def values(self) -> Iterator[_VT]: ...  # type: ignore[override]
weakref.pyi:    def items(self) -> Iterator[tuple[_KT, _VT]]: ...  # type: ignore[override]
webbrowser.pyi:    def open(self, url: str, new: Literal[0, 1, 2] = 0, autoraise: bool = True) -> bool: ...  # type: ignore[override]
xml/dom/expatbuilder.pyi:    def parseFile(self, file: SupportsRead[ReadableBuffer | str]) -> DocumentFragment: ...  # type: ignore[override]
xml/dom/expatbuilder.pyi:    def parseString(self, string: ReadableBuffer | str) -> DocumentFragment: ...  # type: ignore[override]
xml/dom/expatbuilder.pyi:    def parseFile(self, file: SupportsRead[ReadableBuffer | str]) -> None: ...  # type: ignore[override]
xml/dom/expatbuilder.pyi:    def parseString(self, string: str | ReadableBuffer) -> None: ...  # type: ignore[override]
xml/dom/expatbuilder.pyi:    def start_doctype_decl_handler(  # type: ignore[override]
xml/dom/minicompat.pyi:    def __add__(self, other: Iterable[_T]) -> NodeList[_T]: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def insertBefore(  # type: ignore[override]
xml/dom/minidom.pyi:    def appendChild(self, node: _DFChildrenPlusFragment) -> _DFChildrenPlusFragment: ...  # type: ignore[override]
xml/dom/minidom.pyi:    @overload  # type: ignore[override]
xml/dom/minidom.pyi:    def replaceChild(self, newChild: _DocumentFragmentChildren, oldChild: _DFChildrenVar) -> _DFChildrenVar | None: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def removeChild(self, oldChild: _DFChildrenVar) -> _DFChildrenVar: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def insertBefore(self, newChild: _AttrChildrenPlusFragment, refChild: _AttrChildren | None) -> _AttrChildrenPlusFragment: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def appendChild(self, node: _AttrChildrenPlusFragment) -> _AttrChildrenPlusFragment: ...  # type: ignore[override]
xml/dom/minidom.pyi:    @overload  # type: ignore[override]
xml/dom/minidom.pyi:    def replaceChild(self, newChild: _AttrChildren, oldChild: _AttrChildrenVar) -> _AttrChildrenVar | None: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def removeChild(self, oldChild: _AttrChildrenVar) -> _AttrChildrenVar: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def attributes(self) -> NamedNodeMap: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def insertBefore(  # type: ignore[override]
xml/dom/minidom.pyi:    def appendChild(self, node: _ElementChildrenPlusFragment) -> _ElementChildrenPlusFragment: ...  # type: ignore[override]
xml/dom/minidom.pyi:    @overload  # type: ignore[override]
xml/dom/minidom.pyi:    def replaceChild(self, newChild: _ElementChildren, oldChild: _ElementChildrenVar) -> _ElementChildrenVar | None: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def removeChild(self, oldChild: _ElementChildrenVar) -> _ElementChildrenVar: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def appendChild(self, newChild: _EntityChildren) -> Never: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def insertBefore(self, newChild: _EntityChildren, refChild: _EntityChildren | None) -> Never: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def removeChild(self, oldChild: _EntityChildren) -> Never: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def replaceChild(self, newChild: _EntityChildren, oldChild: _EntityChildren) -> Never: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def appendChild(self, node: _DocumentChildrenVar) -> _DocumentChildrenVar: ...  # type: ignore[override]
xml/dom/minidom.pyi:    def removeChild(self, oldChild: _DocumentChildrenVar) -> _DocumentChildrenVar: ...  # type: ignore[override]
xml/dom/minidom.pyi:        self, newChild: _DocumentChildrenPlusFragment, refChild: _DocumentChildren | None  # type: ignore[override]
xml/dom/minidom.pyi:    @overload  # type: ignore[override]
xmlrpc/client.pyi:    def __eq__(self, other: _DateTimeComparable) -> bool: ...  # type: ignore[override]
xmlrpc/server.pyi:    def docroutine(  # type: ignore[override]
zipfile/__init__.pyi:    def readline(self, limit: int = -1) -> bytes: ...  # type: ignore[override]
zipfile/__init__.pyi:    def read1(self, n: int | None) -> bytes: ...  # type: ignore[override]
```

</details>

**Comment by AlexWaygood** [comment]
It's a shame I didn't meet you at EuroPython, it would have been fun to chat about this! 😃

I do agree that this particular LSP violation in the stdlib is (maybe a lot) more impactful than most others. While I still think we should be pretty careful about ignoring typeshed's stubs in cases like this -- there are a _lot_ of LSP violations in the stdlib, and fixing them all with special casing in ty just wouldn't scale -- perhaps you're right that this one is particularly awful.

I'm still not sure if narrow special cases such as the ones you added in your PR are going to ultimately fix the problem, though. The trouble is that the LSP violation means that `date` and `datetime` are fundamentally unsound. Even if we apply special casing, there will still be many ways to upcast a `datetime` to a `date` without ty complaining. Here's the behaviour we get on the PR you opened:

```py
from datetime import date, datetime

date(2020, 1, 1) < datetime(2020, 1, 1, 12, 30)  # error
date(2020, 1, 1) - datetime(2020, 1, 1, 12, 30)  # error


def gimme_dates(x: date, y: date):
    x < y  # no error
    x - y  # no error


gimme_dates(date(2020, 1, 1), datetime(2020, 1, 1, 12, 30))  # no error, raises TypeError at runtime
```

If we want to _actually_ fix the problem, the special casing needs to go much deeper (making `datetime` not actually be considered assignable to `date` at all, as a special case). Otherwise I think we just end up with confusing inconsistencies in our behaviour, and don't make anything significantly safer.

Honestly, whenever I think about this problem, I just end up arriving back at the conclusion that the design of the `datetime` module is fundamentally unsound, and unsuitable for usage by folks who care about their code being strongly typed. So I still feel like I lean towards simply recommending that users use a newer alternative such as `whenever` if they want to have sound datetimes, rather than adding any special casing here.

**Comment by carljm** [comment]
Even if we achieved fully consistent "datetime is not a subtype of date" behavior (and the simplest, least bug-prone way to do that would be to apply https://github.com/python/typeshed/pull/15136 to our vendored typeshed), we would still have problems with runtime narrowing:

```py
def f(obj: object):
    if isinstance(obj, date):
        obj < date(2020, 1, 1)  # we consider `obj` a `date` now, but at runtime it might be a `datetime`, this could error
```

**Comment by carljm** [comment]
I guess maybe we could then special-case `isinstance` narrowing on `date` so that it instead narrows to `date | datetime`, requiring users to also do a `not isinstance(x, datetime)` check to actually get a sound typing as `date`...

And then all code in the wild with `date` annotations that actually wants to accept both `date` and `datetime` has to change the annotation to `date | datetime`.

**Comment by AlexWaygood** [comment]
> And then all code in the wild with `date` annotations that actually wants to accept both `date` and `datetime`

Though the premise of applying that patch to our vendored typeshed stubs would really be that nobody _actually_ ever wants that, even if they think they do...
